src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -795,203 +789,187 @@ where
-                        1000 * 60 * 24 * 3,
-                    ));
-                    if PartialOrder::less_than(&last_compaction_since, &compaction_since) {
-                        trace!(
-                            "remap({id}) {worker_id}/{worker_count}: compacting remap \
-                        shard to: {:?}",
-                            compaction_since,
-                        );
-                        timestamper.compact(compaction_since.clone()).await;
-                        last_compaction_since = compaction_since;
+                    // Note that we choose to compact to WELL before the resumption
+                    // frontier, to paper over some subtle `as_of` < `since` bugs we
+                    // are seeing.
+                    // TODO(guswynn|aljoscha): figure out what is going on here.
+                    //
+                    // Also note this can happen BEFORE we inspect the input. This is somewhat
+                    // of an oddity in the timely world, but this frontier can ONLY advance
+                    // past the input AFTER the output capability of this operator itself has
+                    // been downgraded (which drives the data shard upper), AND the
+                    // remap shard has been advanced below.
+                    //
+                    // Additionally, the empty
+                    // `resumption_frontier` currently occurs only when a source is terminating,
+                    // and we avoid compacting when this happens, as reclocking does not support
+                    // empty frontiers.
+                    //
+                    // TODO(guswynn|petrosagg): support compaction to the empty frontier.
+                    if let Some(upper_ts) = frontier.into_option() {
+                        let compaction_since = Antichain::from_elem(upper_ts.saturating_sub(
+                            // 3 days in milliseconds
+                            1000 * 60 * 24 * 3,
+                        ));
+                        if PartialOrder::less_than(&last_compaction_since, &compaction_since) {
+                            trace!(
+                                "remap({id}) {worker_id}/{worker_count}: compacting remap \
+                            shard to: {:?}",
+                                compaction_since,
+                            );
+                            timestamper.compact(compaction_since.clone()).await;
+                            last_compaction_since = compaction_since;
+                        }
                     }
                 }
-
-                input.for_each(|_cap, data| {
-                    data.swap(&mut buffer);
-
-                    for batch_upper_summary in buffer.drain(..) {
-                        for (pid, offset) in batch_upper_summary.batch_upper.iter() {
-                            global_source_upper.maybe_insert(pid.clone(), *offset);
+                Some(event) = input.next() => match event {
```

Comment:
technically this is slower than `for_each`...maybe add a comment that we should implement a `ForEach` future for `AsyncInputHandle`'s?

Reply:
talked offline, this is actually about the same speed